frontend/catalyst/pennylane_extensions.py

```diff
@@ -1659,3 +1767,102 @@ def _callee():
         return _call_handler(_callee=_callee)
     else:
         raise ValueError(f"Expected a callable or a qml.Operator, not {f}")
+
+
+def ctrl(
+    f: Union[Callable, Operator],
+    control: List[Any],
+    control_values: Optional[List[Any]] = None,
+    work_wires: Optional[List[Any]] = None,
+) -> Callable:
+    """Create a method that applies a controlled version of the provided op. This function is the
+    Catalyst version of the ``qml.ctrl`` that supports Catalyst hybrid operations such as loops and
+    conditionals.
+
+    Args:
+        f (Callable or Operator): A PennyLane operation or a Python function
+                                  containing PennyLane quantum operations.
+        control (Wires): The control wire(s).
+        control_values (List[bool], optional): The value(s) the control wire(s) should take.
+            Integers other than 0 or 1 will be treated as ``int(bool(x))``.
+        work_wires (Any): Any auxiliary wires that can be used in the decomposition
+
+    Returns:
+        (function or :class:`~.operation.Operator`): If an Operator is provided, returns a
+        Controlled version of the Operator.  If a function is provided, returns a function with the
+        same call signature that creates a controlled version of the provided function.
+
+    Raises:
+        ValueError: invalid parameter values, measurements are among the controlled operations.
+
+    **Example**
+
+    .. code-block:: python
+
+        @qjit
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def workflow(theta, w, cw):
+            qml.Hadamard(wires=[0])
+            qml.Hadamard(wires=[1])
+            def func():
+                qml.RX(theta, wires=w)
+                qml.RY(theta, wires=w)
+            catalyst.ctrl(func, control=[cw])()
+            catalyst.ctrl(qml.RZ, control=[cw])(theta, wires=w)
+            catalyst.ctrl(qml.RY(theta, wires=w), control=[cw])
+            return qml.probs()
+
+    >>> workflow(jnp.pi/4, 1, 0)
+    array([0.25, 0.25, 0.03661165, 0.46338835])
+    """
+
+    def _tolist(x):
+        return [x] if not isinstance(x, list) else x
+
+    control = _tolist(control)
+    control_values = _tolist(control_values) if control_values is not None else [1] * len(control)
+    if len(control) != len(control_values):
+        raise ValueError(
+            f"Length of the control_values ({len(control_values)}) must be None or equal "
+            f"to the lenght of control ({len(control)})"
+        )
+    work_wires = _tolist(work_wires) if work_wires is not None else None
+
+    def _call_handler(*args, _callee: Callable, **kwargs):
```

Comment:
Not too important, but I wouldn't use underscores in local variable names / function arguments, this is often used to the indicate the variable is unused in Python (e.g. pylint).

Reply:
Here I kept in mind the concept of "semi-private" entities which is also in use in Python. In general I find it more convenient to distinguish local helper functions from the outer functions. But I am ready to change this.
Note that I used this style before, so if we want to change it, it would probably be better to do it in the whole module at once by a separate commit, what do you think?

Reply:
Oh sorry I didn't mean the function itself, just the function arguments / local variables.

Reply:
I see. Yes, in this case there is another reason :smile:.  `args` and `kwargs` in the function are controlled by users. The underscore in `_callee` minimizes the chance of intersection. It would be an error if users apply ctrl to a function with the argument of the same name. As for the local variables, I can not see any, besides the actual `_` variables which are indeed unused.

Reply:
Oh interesting! In this case it's fine then :)